### PR TITLE
ADO: Implement user permissions sync

### PR DESCRIPTION
**Pull request by indradhanush**
Originally created #48041 for early feedback. I've addressed those and created a new PR because it diverged a lot from the original branch.

Major progress on #47165. There's a little bit of follow up work required which would add more code to this PR - the implementation of `ValidationCheck` on the authz provider. I will follow up with a new PR, but the absence of it does not prevent a good configuration of ADO from working.

## Note to reviewers

Not necessarily a big PR feature wise, ~500 lines is test code. But this does touch multiple modules that would make it hard to piece together for anyone not too familiar with the permissions syncing code. I've broken it down into sequential commits that should make sense when viewed in isolation for the most part. Each commit has a description describing the context and what to expect next. So it might be easier to view each commit at a time and then the PR as a whole.

Also note the usage of `authn` and `authz` in the commit messages. 

## Test plan
- Added tests
- Tested locally
  - Part 1: https://www.loom.com/share/1fad17d488c84a60832ff5931cae196d 
  - Part 2: https://www.loom.com/share/f23e1e5cff3f41c38b47856b71c0f5f7
- Builds should pass

### CODE REVIEW

**Comment by indradhanush** [comment]
After chatting with @kopancek and @pjlast on Slack, I've added a fix for handling multiple code host connections. I will add a test for this tomorrow as I need to step out now. Ran out of time. ⌛

File: internal/database/external_services.go [hunk 1]
```diff
@@ -1598,6 +1598,11 @@ WHERE EXISTS(
 func (e *externalServiceStore) recalculateFields(es *types.ExternalService, rawConfig string) error {
 	es.Unrestricted = !envvar.SourcegraphDotComMode() && !gjson.Get(rawConfig, "authorization").Exists()
 
+	enforcePermissions := gjson.Get(rawConfig, "enforcePermissions")
```
Comment (sashaostrikov) [inline internal/database/external_services.go:1601]: `enforcePermissions` should "override" `authorization` setting, right?
just double-checking, because on L1599 we already set `es.Unrestricted` and here we may override it

File: internal/extsvc/azuredevops/client.go [hunk 2]
```diff
@@ -4,20 +4,24 @@ package azuredevops
...
 	"github.com/sourcegraph/sourcegraph/internal/httpcli"
+	"github.com/sourcegraph/sourcegraph/internal/oauthutil"
 	"github.com/sourcegraph/sourcegraph/internal/ratelimit"
 	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"golang.org/x/oauth2"
 )
 
 const (
-	azureDevOpsServicesURL  = "https://dev.azure.com/"
+	AZURE_DEV_OPS_API_URL   = "https://dev.azure.com/"
```
Comment (sashaostrikov) [inline internal/extsvc/azuredevops/client.go:24]: should we change the case for 2 constants below as well in this case?

File: enterprise/internal/authz/azuredevops/provider_test.go [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			rcache.SetupForTest(t)
+
+			if tc.setup != nil {
+				tc.setup()
+			}
+
+			if tc.teardown != nil {
+				defer tc.teardown()
```
Comment (sashaostrikov) [inline enterprise/internal/authz/azuredevops/provider_test.go:465]:
```suggestion
				t.Cleanup(tc.teardown)
```

File: internal/extsvc/azuredevops/types.go [hunk 4]
```diff
@@ -274,8 +275,21 @@ type CreatorInfo struct {
...
 
-type httpError struct {
+type HTTPError struct {
 	StatusCode int
 	URL        *url.URL
 	Body       []byte
 }
+
+func (e *HTTPError) Error() string {
+	// Do not write HTTPError.Body as part of the error string as the ADO API returns raw HTML for
```
Comment (sashaostrikov) [inline internal/extsvc/azuredevops/types.go:285]: in this case I really prefer this comment to be moved to the doc of the function because this function is a one-liner and you basically explain what this function does and not a particular small part of its code, right?

File: internal/extsvc/azuredevops/users.go [hunk 5]
```diff
@@ -12,7 +12,7 @@ import (
 	"golang.org/x/oauth2"
 )
 
-const VISUAL_STUDIO_APP_URL = "https://app.vssps.visualstudio.com"
+const VISUAL_STUDIO_APP_URL = "https://app.vssps.visualstudio.com/"
```
Comment (sashaostrikov) [inline internal/extsvc/azuredevops/users.go:15]: won't this break anything unrelated to changes of this PR?
just a thing to double-check :)

File: internal/oauthutil/oauth2.go [hunk 6]
```diff
@@ -27,6 +27,9 @@ type OAuthContext struct {
 	Endpoint Endpoint
 	// Scope specifies optional requested permissions.
 	Scopes []string
+
+	// CustomURLArgs if any needed by a specific provider that may have a custom implementation.
```
Comment (sashaostrikov) [inline internal/oauthutil/oauth2.go:31]: I cannot get the meaning of the comment, maybe it can be changed a bit?
```suggestion
	// CustomURLArgs are URL parameters which may be needed by a specific provider that may have a custom OAuth2 implementation.
```

File: internal/oauthutil/token.go [hunk 7]
```diff
@@ -117,6 +117,20 @@ func newTokenRequest(oauthCtx OAuthContext, refreshToken string, authStyle AuthS
 		v.Set("refresh_token", refreshToken)
 	}
 
+	// FUTURE REFACTORING NOTE: Most of the code in this module is very similar to the upstream
```
Comment (sashaostrikov) [inline internal/oauthutil/token.go:120]: `TODO` probably?

**Review by sashaostrikov — COMMENTED** [review]
Great work! Left a bunch of comments, but nothing is critical, mostly related to how the code is written and not about the correctness of implementation :)

File: internal/extsvc/azuredevops/users.go (hunk 5 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/users.go:15]: Yes, I've checked it - the full OAuth2 flow still works (which is where this is used). I've updated the places where this is used and I have tests for this which continue to pass. Good catch though. 👍🏽

File: internal/extsvc/azuredevops/client.go [hunk 8]
```diff
@@ -97,25 +101,28 @@ func (c *client) do(ctx context.Context, req *http.Request, urlOverride string,
...
 	}
 
-	resp, err := c.httpClient.Do(req)
+	logger := log.Scoped("azuredevops.Client", "azuredevops Client logger")
+	resp, err := oauthutil.DoRequest(ctx, logger, c.httpClient, req, c.auth)
 	if err != nil {
 		return "", err
 	}
 
+	defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go:117]: Great catch. 👍

File: internal/database/external_services.go (hunk 1 above)
Comment (indradhanush) [reply internal/database/external_services.go:1601]: Yes. `authorization` is still the relevant way of doing it for all codes hosts except Azure DevOps. So only override `es.Unrestricted`, if `enforcePermissions` is set. 

I will add a comment here.

File: enterprise/internal/authz/azuredevops/provider.go [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+		urn:      "azuredevops:authzprovider",
+		codeHost: extsvc.NewCodeHost(u, extsvc.TypeAzureDevOps),
+		orgs:     orgs,
+		projects: projects,
+	}, nil
+}
+
+func NewAuthzProviders(db database.DB, conns []*types.AzureDevOpsConnection) *authztypes.ProviderInitResult {
+	initResults := &authztypes.ProviderInitResult{}
+	orgs, projects := []string{}, []string{}
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:225]: I find the current approach simpler. Especially since the final length of orgs and projects is not known beforehand so we cannot use that information to create a N capacity slice ahead of time.

File: enterprise/internal/authz/azuredevops/provider.go [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+	return &Provider{
+		db:       db,
+		urn:      "azuredevops:authzprovider",
+		codeHost: extsvc.NewCodeHost(u, extsvc.TypeAzureDevOps),
+		orgs:     orgs,
+		projects: projects,
+	}, nil
+}
+
+func NewAuthzProviders(db database.DB, conns []*types.AzureDevOpsConnection) *authztypes.ProviderInitResult {
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:224]: Makes sense. 👌

File: enterprise/internal/authz/azuredevops/provider.go [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+func (p *Provider) ServiceID() string {
+	return p.codeHost.ServiceID
+}
+
+func (p *Provider) URN() string {
+	return p.urn
+}
+
+// TODO: Implement this in a follow up PR.
+func (p *Provider) ValidateConnection(ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:201]: Going to implement this in a follow PR next. Will leave it as it is.

File: enterprise/internal/authz/azuredevops/provider.go [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+				// have access to the API returns 401. We're not sure if the API might return 403
+				// for some use case but we don't want to hard fail on that either.
+				if httpErr.StatusCode == http.StatusUnauthorized || httpErr.StatusCode == http.StatusForbidden || httpErr.StatusCode == http.StatusNotFound {
+
+					l.Debug("skipping org",
+						log.String("org", org),
+						log.Int("http status code", httpErr.StatusCode),
+					)
+
+					continue
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:122]: Hmm. I wouldn't log it as a `warn` since that will show up on more or less every permissions sync - I'm guessing not all users have access to all orgs / projects. I'm keeping it as a `DEBUG` for times if we have to debug a permissions issue and we want to verify that this `org` or `project` is explicitly being skipped because of a `4xx` error - that is the user does not have access to it.

File: internal/extsvc/azuredevops/client.go (hunk 2 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go:24]: I don't think we should - because they are only used within the package so it's fine for them to be lower case.

File: internal/extsvc/azuredevops/client.go [hunk 13]
```diff
@@ -97,25 +101,28 @@ func (c *client) do(ctx context.Context, req *http.Request, urlOverride string,
 		req.Header.Set("Content-Type", "application/json")
 	}
 
-	// Add Basic Auth headers for authenticated requests.
+	// Add authentication headers for authenticated requests.
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go:104]: Hmm. This is an oversight. Shouldn't really be ignoring this error. Will fix.

File: internal/oauthutil/token.go (hunk 7 above)
Comment (indradhanush) [inline internal/oauthutil/token.go:120]:
```suggestion
	// TODO: (FUTURE REFACTORING NOTE) Most of the code in this module is very similar to the upstream
```

File: enterprise/internal/authz/azuredevops/provider_test.go (hunk 3 above)
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider_test.go]: Haha! My Python muscle memory slithering in here. 😅

File: internal/oauthutil/oauth2.go [hunk 14]
```diff
@@ -27,6 +27,9 @@ type OAuthContext struct {
...
+	CustomURLArgs map[string]string
```
Comment (kopancek) [inline internal/oauthutil/oauth2.go:32]: I would call it CustomQueryParams as that better conveys the meaning to me. We add them to URL as query parameters, so why not call them that?

File: internal/oauthutil/token.go [hunk 15]
```diff
@@ -117,6 +117,20 @@ func newTokenRequest(oauthCtx OAuthContext, refreshToken string, authStyle AuthS
...
+	//
+	// If we decide to refactor this, OAuthContext can be replaced by oauth2.Config which is pretty
+	// much the same except the CustomURLArgs that we're adding here. But that also means if we go
+	// back to using oauth2.Config, we can use the Exchange method instead and pass the custom args
+	// as oauth2.AuthCodeOption values. See the implementation of the azuredevops auth provider
+	// which already does this to exchange the auth_code for an access token the first time a user
+	// connects their ADO account with Sourcegraph.
+	for key, value := range oauthCtx.CustomURLArgs {
+		v.Set(key, value)
+	}
```
Comment (kopancek) [inline internal/oauthutil/token.go:132]: This might theoretically override the params set above, depending on the policy the consumer uses on the query parameters. Just wanted to make sure that we are aware of that.

File: schema/azuredevops.schema.json [hunk 16]
```diff
@@ -19,6 +19,11 @@
       "format": "uri",
       "examples": ["https://dev.azure.com"]
     },
+    "enforcePermissions": {
+      "description": "A flag to enforce Azure DevOps repository access permissions",
+      "type": "boolean",
+      "default": false
+    },
```
Comment (kopancek) [inline schema/azuredevops.schema.json:26]: Please, add docs for this once you start making docs changes. This is a very big step forward, but also completely different to what we configure for other code hosts to enable permissions.

File: internal/database/external_services.go (hunk 1 above)
Comment (mrnugget) [reply internal/database/external_services.go:1612]: > Yes. `authorization` is still the relevant way of doing it for all codes hosts except Azure DevOps

Curious: why is it different for ADO?

File: internal/extsvc/azuredevops/client.go [hunk 17]
```diff
@@ -156,17 +163,81 @@ func (c *client) GetURL() *url.URL {
...
+				},
+				// The API expects some custom values in the POST body to refresh the token. See:
+				// https://learn.microsoft.com/en-us/azure/devops/integrate/get-started/authentication/oauth?view=azure-devops#4-use-the-access-token
+				//
+				// DEBUGGING NOTE: The token refresher (internal/oauthutil/token.go:newTokenRequest)
+				// adds some default key-value pairs to the body, some of which are eventually
+				// overridden by the values in this map. But some extra arg remain in the body that
+				// is sent in the request. This works for now, but if refreshing a token ever stops
+				// working for Azure Dev Ops this is a good place to start looking by writing a
+				// custom implementation that only sends the key-value pairs that the API expects.
```
Comment (kopancek) [inline internal/extsvc/azuredevops/client.go:206]: Since you are aware of this problem, wouldn't it be a better architecture to actually do what you suggest? E.g. add a function on the context that would provide URL query parameters and default to the original one if it's not present?

File: internal/extsvc/azuredevops/client.go (hunk 2 above)
Comment (sashaostrikov) [reply internal/extsvc/azuredevops/client.go:24]: well, okay, that's a fair point.
I found that there is a convention in Go not to use SNAKE_CASE at all, let's fix it, then and use camelCase or PascalCase everywhere depending on being exported or not

https://go.dev/doc/effective_go#mixed-caps

File: enterprise/internal/authz/azuredevops/provider.go [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+
+	var apiURL string
+	if MOCK_API_URL != "" {
+		apiURL = MOCK_API_URL
+	} else {
+		apiURL = azuredevops.AZURE_DEV_OPS_API_URL
+	}
+
+	client, err := azuredevops.NewClient(
+		p.ServiceID()+account.AccountID,
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:128]: Me too. But `gofmt` thinks it knows better. 🤷🏽

File: enterprise/cmd/frontend/internal/auth/azureoauth/session.go [hunk 19]
```diff
@@ -58,7 +58,7 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 		},
 		ExternalAccount: extsvc.AccountSpec{
 			ServiceType: s.ServiceType,
-			ServiceID:   s.ServiceID,
+			ServiceID:   azuredevops.AZURE_DEV_OPS_API_URL,
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/cmd/frontend/internal/auth/azureoauth/session.go:61]: Currently the Azure Dev Ops integration will only work for `dev.azure.com`, which is why I'm making it a constant. The ADO enterprise version uses different APIs so to support that we will need to do extra work. Is that okay?

The reason I had to make this a constant though - 

1. The OAuth API uses a different URL for the authentication flow and that was bleeding in here causing all sorts of bugs with one part of the code expecting dev.azure.com and another saving `app.vssps.visualstudio.com/`. 🤦🏽

File: internal/database/external_services.go (hunk 1 above)
Comment (indradhanush) [reply internal/database/external_services.go:1612]: I'm introducing a new pattern because it felt really hacky to add a `authorization` object on the schema with no fields. For ADO, we don't need any custom fields but just a boolean to enforce the permissions. The change was simple enough that it made sense to go ahead with.

File: enterprise/internal/authz/azuredevops/provider.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go]: Good idea. I changed it to: `user does not have access to this org`. What do you think?

File: internal/extsvc/azuredevops/client.go (hunk 2 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go:24]: Makes sense. Fixed.

File: internal/extsvc/azuredevops/client.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go:104]: Now fixed.

File: internal/oauthutil/token.go (hunk 15 above)
Comment (indradhanush) [reply internal/oauthutil/token.go:132]: Yes, 100%. This _will_ override the params set above - which is the right use case that this customisation is aimed towards. To override default params and add custom ones.

File: enterprise/internal/authz/azuredevops/provider.go [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+
+		logger.Debug("adding repos", log.Int("count", len(foundRepos)))
+		repos = append(repos, foundRepos...)
+	}
+
+	seenProjects := map[string]struct{}{}
+	for _, project := range p.projects {
+		// The list of projects may have duplicates if there are multiple Azure DevOps code host
+		// connections that have the same project in their config. As a result, skip this project if
+		// we already listed its repositories in this iteration of permissions syncing.
```
Comment (pjlast) [inline enterprise/internal/authz/azuredevops/provider.go:191]: Similarly, rather do this once when creating the provider than on every sync.

File: internal/oauthutil/oauth2.go (hunk 14 above)
Comment (indradhanush) [reply internal/oauthutil/oauth2.go]: That's a great name. I really struggled with a good name for this one. Thank you. Fixed.

File: enterprise/internal/authz/azuredevops/provider.go [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+
+func newAuthzProvider(db database.DB, orgs, projects []string) (*Provider, error) {
+
+	if err := licensing.Check(licensing.FeatureACLs); err != nil {
+		return nil, err
+	}
+
+	u, err := url.Parse(azuredevops.AZURE_DEV_OPS_API_URL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse url: %q, this is likely a misconfigured URL in the constant azuredevops.AZURE_DEV_OPS_API_URL", azuredevops.AZURE_DEV_OPS_API_URL)
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go]: The constant is defined in code. I don't think users can fix it - mostly a quick hint for devs on where to look if this ever gets reported to us as a customer issue.

File: internal/extsvc/azuredevops/client.go [hunk 22]
```diff
@@ -156,17 +163,81 @@ func (c *client) GetURL() *url.URL {
 // IsAzureDevOpsServices returns true if the client is configured to Azure DevOps
 // Services (https://dev.azure.com
 func (c *client) IsAzureDevOpsServices() bool {
-	return c.URL.String() == azureDevOpsServicesURL
+	return c.URL.String() == AZURE_DEV_OPS_API_URL
+}
+
+func GetOAuthContext(refreshToken string) (*oauthutil.OAuthContext, error) {
+	for _, authProvider := range conf.SiteConfig().AuthProviders {
+		if authProvider.AzureDevOps != nil {
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go:178]: You cannot have more than one auth provider for Azure at the moment. We don't want to allow that - since we only support `dev.azure.com` at the moment. Is there a reason why we would want to support that for the same URL?

File: enterprise/internal/authz/azuredevops/provider.go [hunk 23]
```diff
@@ -0,0 +1,256 @@
...
+		)
+	}
+
+	var repos []azuredevops.Repository
+
+	seenOrgs := map[string]struct{}{}
+	for _, org := range p.orgs {
+		// The list of orgs may have duplicates if there are multiple Azure DevOps code host
+		// connections that have the same org in their config. As a result, skip this org if we
+		// already listed its repositories in this iteration of permissions syncing.
```
[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:166]: Great catch. Fixed.

File: internal/extsvc/azuredevops/client.go [hunk 24]
```diff
@@ -97,25 +101,34 @@ func (c *client) do(ctx context.Context, req *http.Request, urlOverride string,
...
+	logger := log.Scoped("azuredevops.Client", "azuredevops Client logger")
+	resp, err := oauthutil.DoRequest(ctx, logger, c.httpClient, req, c.auth)
 	if err != nil {
 		return "", err
 	}
 
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			logger.Error("failed to close body", log.String("err", err.Error()))
+		}
```
[the remark this answers is not in the document]
Comment (pjlast) [reply internal/extsvc/azuredevops/client.go:123]: I'm not convinced 🙈 even if there were a spurt in these errors, we still can't do anything about them, and they don't affect the functioning of the program at all. The best they can do is make someone worry about errors in the logs that don't mean anything.

This is also what the Go docs say: https://sourcegraph.com/github.com/golang/go/-/blob/src/net/http/doc.go?L19-25&subtree=true

And a bunch of examples from the standard lib as well: https://sourcegraph.com/search?q=context%3Aglobal+repo%3A%5Egithub%5C.com%2Fgolang%2Fgo%24+defer+resp.Body.Close%28%29&patternType=standard&sm=1&groupBy=path

There is one example in the standard lib where they check for the err, and it's in a test to assert that the error should actually have been `nil` https://sourcegraph.com/github.com/golang/go/-/blob/src/net/http/client_test.go?L2028:11&subtree=true

[the remark this answers is not in the document]
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go]: Thank you for the references. I'm convinced now. 😁

File: internal/extsvc/azuredevops/client.go (hunk 17 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go:205]: If by context you mean `OAuthContext` here, then from what I understand you're suggesting:

1. Add a method on `OAuthContext` like: `func (o *OAuthContext) AddExtraArgs()`
1. Call that instead of reading from `CustomQueryParams`?

If it's not a blocker, I would like to keep it as is - for now, since this PR is big enough and because we do a very similar workflow in the OAuth login by passing in extra args as part of the `Exchange` method call.

File: enterprise/internal/authz/azuredevops/provider.go [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	u, err := url.Parse(azuredevops.AzureDevOpsApiUrl)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse url: %q, this is likely a misconfigured URL in the constant azuredevops.AzureDevOpsApiUrl", azuredevops.AzureDevOpsApiUrl)
+	}
+
+	return &Provider{
+		db:       db,
+		urn:      "azuredevops:authzprovider",
```
Comment (indradhanush) [inline enterprise/internal/authz/azuredevops/provider.go:83]: @pjlast Because you mentioned about the URN in the client in another comment, I hope this one is not an issue? In the other providers we use the connection.URN, but in this case since we create one provider for all connections, I thought maybe it's okay to have a hardcoded string on the Provider?

File: internal/extsvc/azuredevops/client.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply internal/extsvc/azuredevops/client.go]: now I understand too why we don't catch them in our code base :)

**Review by sashaostrikov — APPROVED** [review]
At least all of my comments are resolved -- LGTM :shipit:

File: enterprise/internal/authz/azuredevops/provider.go [hunk 26]
```diff
@@ -0,0 +1,248 @@
...
+	// Nothing fancy, just some data restructuring to convert the map back to a slice now that we
+	// have a unique list of orgs and projects.
+	uniqueOrgs, uniqueProjects := []string{}, []string{}
+	for name := range orgs {
+		uniqueOrgs = append(uniqueOrgs, name)
+	}
+
+	for name := range projects {
+		uniqueProjects = append(uniqueProjects, name)
+	}
```
Comment (kopancek) [inline enterprise/internal/authz/azuredevops/provider.go:58]: You can now do this with [`maps.Keys`](https://pkg.go.dev/golang.org/x/exp/maps#Keys). Which is more convenient than writing a for loop :)
```suggestion
	// Convert the map back to a slice now that we have a unique list of orgs and projects.
	uniqueOrgs := maps.Keys(orgs)
	uniqueProjects := maps.Keys(projects)
```

File: enterprise/internal/authz/azuredevops/provider.go (hunk 25 above)
Comment (kopancek) [reply enterprise/internal/authz/azuredevops/provider.go:83]: This is important to get right, especially for the user permission sync? But I'm confused right now what is right or wrong here 😅

File: enterprise/internal/authz/azuredevops/provider.go [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+		extIDs = append(extIDs, extsvc.RepoID(repo.ID))
+	}
+
+	return &authz.ExternalUserPermissions{
+		Exacts: extIDs,
+	}, err
+}
+
+func (p *Provider) FetchRepoPerms(ctx context.Context, repo *extsvc.Repository, opts authz.FetchPermsOptions) ([]extsvc.AccountID, error) {
+	return nil, authz.ErrUnimplemented{Feature: "azuredevops.FetchRepoPerms"}
```
[the remark this answers is not in the document]
Comment (kopancek) [reply enterprise/internal/authz/azuredevops/provider.go:230]: Can you add a comment about it to the code?

File: internal/extsvc/azuredevops/client.go [hunk 28]
```diff
@@ -156,17 +163,81 @@ func (c *client) GetURL() *url.URL {
 // IsAzureDevOpsServices returns true if the client is configured to Azure DevOps
 // Services (https://dev.azure.com
 func (c *client) IsAzureDevOpsServices() bool {
-	return c.URL.String() == azureDevOpsServicesURL
+	return c.URL.String() == AZURE_DEV_OPS_API_URL
+}
+
+func GetOAuthContext(refreshToken string) (*oauthutil.OAuthContext, error) {
+	for _, authProvider := range conf.SiteConfig().AuthProviders {
```
[the remark this answers is not in the document]
Comment (kopancek) [reply internal/extsvc/azuredevops/client.go:173]: I still think the above ☝️ would be a nice improvement to the code.

File: internal/extsvc/azuredevops/client.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply internal/extsvc/azuredevops/client.go:174]: I didn't know about this limitation. It should probably be fine.

File: internal/extsvc/azuredevops/client.go (hunk 17 above)
Comment (kopancek) [reply internal/extsvc/azuredevops/client.go:205]: Yes, that's what I had in mind. Not a blocker, just a suggestion for future

**Review by kopancek — APPROVED** [review]
Approving to unblock, but left some comments for future PR maybe :)

File: internal/extsvc/azuredevops/client.go (hunk 17 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go:205]: Noted. I can follow up once this is merged. 👍

File: schema/azuredevops.schema.json (hunk 16 above)
Comment (pjlast) [inline schema/azuredevops.schema.json:26]: GitHub won't allow me to comment further down, so I'm going to leave the comment here.
We can consider making the "what repos do I sync" part of the config a bit more expressive. Here's the scenario as I see it:
- There are organizations, projects, and repositories
- If you want to sync all repositories from an organization, you only specify that organization. It is redundant to be able to specify `org/project` if you've already specified `org`
- Ditto for `org/project/repository`
- So, construct the object in such a way that we get an object as follows:

```json
"repositories": [
  {
    "org": "orgName", // required
    "projects": [ // optional
      {
        "project": "projectName", // required
        "repositoryIds": [ "1", "2" ] // optional
      }
    ]
  }
]
```

So now you cannot specify redundant fields. You either specify an entire org, or you specify an org and some projects, or you specify an org and some projects and repo IDs for some projects. Or any combination of these.

Not a blocker, but I think "ergonomically" it makes a lot of sense.

**Review by pjlast — APPROVED** [review]

File: enterprise/internal/authz/azuredevops/provider.go (hunk 26 above)
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:58]: This is nice. TIL.

[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go]: Hmm. I accepted @kopancek's suggestion to use map.Keys and it looks a lot cleaner in its current form. Will go ahead with that.

File: schema/azuredevops.schema.json (hunk 16 above)
Comment (indradhanush) [reply schema/azuredevops.schema.json:26]: This is a good callout but I'm also not sure if the propsed structure would be less confusing since we nest org and projects under `repositories` - which feels wrong hierarchically since repos are part of an org and not the other way round.

We could however make our backend smarter to ignore projects if orgs have already been synced. 

Let's make this a separate discussion and get others opinions.

Comment (pjlast) [reply schema/azuredevops.schema.json:26]: I mean I just called it repositories to give it a name. It's just a JSON key you can name it anything

Comment (pjlast) [reply schema/azuredevops.schema.json:26]: But yes, splitting the projects by org and dropping redundant ones is the other option

Comment (mrnugget) [reply schema/azuredevops.schema.json:26]: So this replaces the `authorization: {}` thing from the other code host connections, yes?

It kinda makes me nervous that we do this only for one new provider now. Somehow I'd rather have us introduce to every provider at the same time and keep backwards compatibility. But I guess we can still do that.

Comment (indradhanush) [reply schema/azuredevops.schema.json:26]: I agree why it could make anyone nervous. When I was looking at this - the change turned out to be reasonably simple and easy to implement: https://github.com/sourcegraph/sourcegraph/pull/48209/files#diff-e64c2422ddc2f4da136695fcb608eee988e39ac2fb9ed1975646bf3f9bba7766R1612-R1615

Which gave me a high degree of confidence that this will not break for existing code hosts on the `authorization: {}` pattern. And it felt weird to continue to add on to this pattern as a result since the end user site-admin experience of configuring Azure DevOps with permisisons sync wouldn't be very intuitive.

Release timing wise this might be a better time to introduce this pattern limited to Azure DevOps and then refactor the existing ones on the new pattern after the upcoming release so we have sufficient time with testing and knowing if anything has gone wrong until the next release.

File: enterprise/internal/authz/azuredevops/provider.go [hunk 29]
```diff
@@ -0,0 +1,252 @@
...
+		// Add them to a map so that we may filter out duplicates before passing them over to the
+		// provider.
+		for _, name := range c.Orgs {
+			orgs[name] = struct{}{}
+		}
+
+		for _, name := range c.Projects {
+			projects[name] = struct{}{}
+		}
+	}
```
Comment (pjlast) [inline enterprise/internal/authz/azuredevops/provider.go:48]: One thing I forgot to mention here: you need to make sure that you are only doing this for connections with `authorization` enabled, otherwise perms syncing will possibly be fetching a whole lot of redundant orgs and projects.

File: enterprise/internal/authz/azuredevops/provider.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (pjlast) [reply enterprise/internal/authz/azuredevops/provider.go]: I guess 🤷‍♂️ although it looks "cleaner", using `map.Keys` still results in code that loops twice

File: enterprise/internal/authz/azuredevops/provider.go (hunk 29 above)
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:48]: Just saw this. Will fix in a follow up.

File: enterprise/internal/authz/azuredevops/provider.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (kopancek) [reply enterprise/internal/authz/azuredevops/provider.go]: @indradhanush See my updated comment, for usage of map as a set, `maps.Keys` do not work properly 😭

Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go]: @kopancek Can you please elaborate what you mean by does not work as expected?

[the remark this answers is not in the document]
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go]: Hmm. Replied.

Comment (kopancek) [reply enterprise/internal/authz/azuredevops/provider.go]: Nevermind, it does work. The problem was between my chair and my screen, as usual 🤦

Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go]: Ah. The famed Operator Error. 😛

File: internal/extsvc/azuredevops/client.go (hunk 2 above)
Comment (mrnugget) [reply internal/extsvc/azuredevops/client.go]: Looks like we're not super consistent with the `CamelCase`, `SNAKE_CASE` ourselves: https://sourcegraph.sourcegraph.com/search?q=context:sourcegraph+const%5Cs%5BA-Z%5D%2B+lang:go&patternType=regexp&sm=0&groupBy=repo

File: enterprise/internal/authz/azuredevops/provider.go (hunk 29 above)
Comment (indradhanush) [reply enterprise/internal/authz/azuredevops/provider.go:48]: Follow up here: https://github.com/sourcegraph/sourcegraph/pull/48209#discussion_r1122990998

File: internal/extsvc/azuredevops/client.go (hunk 2 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/client.go]: Personally I've always preferred using `SNAKE_CASE` for constants - because it's a subtle hint when reading / writing code that it's a constant which is not something I can derive from `CamelCase` without looking at its type. This is probably one of the few Go conventions I'm not a huge fan of. Add to that variable names like `HTTPAPIURL` become awful to read because you want to use `UPPERCASE` for well-known abbreviations. 🤷🏽‍♂️

Just my two cents. If this was up for a vote in our codebase to align one way or another, I would vote for `SNAKE_CASE` for constants. 😛
